lib/srv/db/common/reporting.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+type reporterConfig struct {
+	engine    Engine
+	clock     clockwork.Clock
+	component string
+	database  types.Database
+}
+
+func (r *reporterConfig) CheckAndSetDefaults() error {
+	if r.engine == nil {
+		return trace.BadParameter("missing parameter engine")
+	}
+	if r.database == nil {
+		return trace.BadParameter("missing parameter database")
+	}
+	if r.clock == nil {
+		r.clock = clockwork.NewRealClock()
+	}
+	if r.component == "" {
+		r.component = teleport.ComponentDatabase
+	}
+	return nil
+}
+
+type reportingEngine struct {
+	reporterConfig
+}
+
+func init() {
+	_ = metrics.RegisterPrometheusCollectors(prometheusCollectorsEngine...)
+}
+
+func NewReportingEngine(cfg reporterConfig) (Engine, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &reportingEngine{cfg}, nil
+}
+
+func (r *reportingEngine) InitializeConnection(clientConn net.Conn, sessionCtx *Session) error {
+	closer := utils.NewCloserConn(clientConn)
+	trackingClientConn := utils.NewTrackingConn(closer)
+
+	go func() {
+		var txPrev, rxPrev uint64
+
+		update := func() {
+			tx, rx := trackingClientConn.Stat()
+			clientWrittenBytes.With(r.getLabels()).Add(float64(tx - txPrev))
+			clientReadBytes.With(r.getLabels()).Add(float64(rx - rxPrev))
+			txPrev = tx
+			rxPrev = rx
+		}
+
+		for {
+			select {
+			case <-closer.Context().Done():
+				update()
+				return
+			case <-r.clock.After(time.Second):
+				update()
+			}
+		}
+	}()
+
+	initializedConnections.With(r.getLabels()).Inc()
+	return r.engine.InitializeConnection(trackingClientConn, sessionCtx)
+}
+
+func (r *reportingEngine) SendError(err error) {
+	engineErrors.With(r.getLabels()).Inc()
+	r.engine.SendError(err)
+}
+
+func (r *reportingEngine) HandleConnection(ctx context.Context, sessionCtx *Session) error {
+	activeConnections.With(r.getLabels()).Inc()
+	defer activeConnections.With(r.getLabels()).Dec()
+
+	start := r.clock.Now()
+	defer func() {
+		connectionDurations.With(r.getLabels()).Observe(r.clock.Since(start).Seconds())
+	}()
+
+	return trace.Wrap(r.engine.HandleConnection(ctx, sessionCtx))
+}
+
+var _ Engine = &reportingEngine{}
+
+var commonLabels = []string{teleport.ComponentLabel, "db_protocol", "db_type"}
+
+func getLabels(component string, db types.Database) prometheus.Labels {
+	return map[string]string{
+		teleport.ComponentLabel: component,
+		"db_protocol":           db.GetProtocol(),
+		"db_type":               db.GetType(),
+	}
+}
+
+func (r *reportingEngine) getLabels() prometheus.Labels {
+	return getLabels(r.component, r.database)
+}
+
+var (
+	clientReadBytes = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: teleport.MetricNamespace,
+			Name:      "db_client_read_bytes_total",
+			Help:      "Total number bytes read from the DB clients",
+		},
+		commonLabels,
+	)
+
+	clientWrittenBytes = prometheus.NewCounterVec(
```

Comment:
Could you also add these metrics to our metrics docs: https://github.com/gravitational/teleport/blob/871aa2af1b953e4d18e528b1bef2ae324efa2e6d/docs/pages/includes/metrics.mdx#L107

Reply:
Yes, I'll raise a separate docs-only PR once this one lands (and metric names are settled etc.).